### PR TITLE
chore(compiler): add comments and update types for in-memory FS

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
This updates the in-memory FS file with a bunch of comments explaining
what various methods on the `InMemoryFileSystem` object do.

It also refactors where the relevant types live. Instead of having a
separate interface declared outside of the file where the
implementation is, we instead derive the `InMemoryFileSystem` from the
return value of the function which creates the object. This lets us have
comments annotating the implementation code which will also show up as
type hints throughout the codebase.

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Unit tests (`npm test`) were run locally and passed
- [x] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [x] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe):

## What is the new behavior?

This does not (or should not!) change how anything works, but is instead just a documentation / cleanup patch.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

I confirmed that unit tests are passing and whatnot, and did a bit of smoke testing with some test components locally.

### CODE REVIEW

File: src/compiler/build/compiler-ctx.ts
Comment (alicewriteswrongs) [inline src/compiler/build/compiler-ctx.ts:5]: the `InMemoryFileSystem` type is now defined in the same file as `createInMemoryFileSystem` and is also exported from `src/compiler/index.ts`. I updated imports to use relative imports _within_ the `compiler` module (i.e. in any file below `src/compiler/`) and I imported from `@stencil/core/compiler` _outside_ of the module, for the sake of consistency and, well, modularity 😄

File: src/compiler/sys/in-memory-fs.ts
Comment (alicewriteswrongs) [inline src/compiler/sys/in-memory-fs.ts:37]: this type is now defined in this file, and is derived from the return type of `createInMemoryFs` instead of being declared separately as an interface. I think this just helps us keep things in one place.

File: src/compiler/sys/in-memory-fs.ts
Comment (alicewriteswrongs) [inline src/compiler/sys/in-memory-fs.ts:44]: some of these other interfaces are used outside of this file (although generally not too often). I considered moving them all to the bottom of the file below `createInMemoryFs` but didn't so yet. If that's a cleaner look I'm open to it!

File: src/compiler/sys/in-memory-fs.ts
Comment (alicewriteswrongs) [inline src/compiler/sys/in-memory-fs.ts:154]: As always, and perhaps especially with this file, these comments reflect what I learned from reading the code, looking for call sites across the codebase, and a little bit of github / slack spelunking here and there. So there very well may be inaccuracies!

File: src/compiler/sys/in-memory-fs.ts
Comment (alicewriteswrongs) [inline src/compiler/sys/in-memory-fs.ts:162]: a note here - the `accessData` function wasn't called outside of this file, and was just sort of odd. I refactored this `access` function to work more like the `accessSync` function below and was able to delete `accessData`

File: src/compiler/sys/in-memory-fs.ts
Comment (alicewriteswrongs) [inline src/compiler/sys/in-memory-fs.ts:724]: I could find no references to these functions in the codebase, so I think they should be safe to delete

**Review by alicewriteswrongs — COMMENTED** [review]
just pointing out a few things

File: src/compiler/sys/in-memory-fs.ts
Comment (rwaskiewicz) [inline src/compiler/sys/in-memory-fs.ts:209]: No action required - IMO we can elide `@returns` for empty Promises from JSDoc. The linter is configured to permit that, but I don't believe we could configure it to enforce consistency (I.E. "warn us if we add `@returns` on a fn that returns `void` or `Promise<void>`), hence the no change needed here

File: src/compiler/sys/in-memory-fs.ts
Comment (rwaskiewicz) [inline src/compiler/sys/in-memory-fs.ts:487]:
```suggestion
   * This is done by marking it for deletion. The item will remain in memory
```

File: src/compiler/sys/in-memory-fs.ts
Comment (rwaskiewicz) [inline src/compiler/sys/in-memory-fs.ts:505]:
```suggestion
   * Remove a directory from the in-memory FS by marking it to be deleted
```

File: src/compiler/sys/in-memory-fs.ts
Comment (rwaskiewicz) [inline src/compiler/sys/in-memory-fs.ts:535]:
```suggestion
   * Remove an item from the in-memory FS by marking it to be deleted
```

File: src/compiler/index.ts
Comment (rwaskiewicz) [inline src/compiler/index.ts:25]: Should we be exporting `InMemoryFileSystem`? Previously that was a private type. It's not that we can't export it, but it does expand the scope of things that are public facing that we maintain

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]

File: .github/workflows/test-browserstack.yml
Comment (alicewriteswrongs) [reply .github/workflows/test-browserstack.yml]: it was failing before and I was trying to get it to print what changed haha

File: src/compiler/sys/in-memory-fs.ts
Comment (alicewriteswrongs) [reply src/compiler/sys/in-memory-fs.ts:37]: My thought was mainly just that it cuts down on verbosity and lets us have comments for these functions that live right next to their implementations, previously there were some comments in the actual implementation, some in the interface, etc. I did port over those 'Synchronous!! ...' comments to the new JSDocs for all those functions so we still have those around, not sure if that addresses what you're thinking about?

File: src/compiler/sys/in-memory-fs.ts
Comment (alicewriteswrongs) [reply src/compiler/sys/in-memory-fs.ts:35]: actually scratch that — it's been a while so I forgot, but the reason that I removed this code (`accessData`) was because it calls the `stat` function and, afaict, the assignments which modify the `item` returned by `getItem` are redundant with what is done there, see:

https://github.com/ionic-team/stencil/blob/2e773244f43cee77febb1715d1393bfa4df1309f/src/compiler/sys/in-memory-fs.ts#L558-L589

there we already have a series of checks which ensure that `.isFile`, `.isDirectory`, and so on are properly set

File: src/compiler/index.ts
Comment (alicewriteswrongs) [reply src/compiler/index.ts:25]: yeah it's not necessary I think, tbh I don't remember why I added it here so I'll remove it

**Comment by alicewriteswrongs** [comment]
@rwaskiewicz I think all comments should be addressed now and (I think...) your changes for that git context stuff should get the build passing now

File: src/compiler/sys/in-memory-fs.ts
Comment (rwaskiewicz) [reply src/compiler/sys/in-memory-fs.ts:37]: Yeah, I think that works!

File: src/compiler/index.ts
Comment (rwaskiewicz) [inline src/compiler/index.ts]: Is `FsWriteResults` re-exported anywhere else? The file that originally had that type definition `stencil-public-compiler` is a signal that this was a publicly exported type - I think we need to continue to re-export it to not break our API (unless I'm missing where it's being re-exported)

File: src/compiler/sys/in-memory-fs.ts
Comment (rwaskiewicz) [reply src/compiler/sys/in-memory-fs.ts:35]: Ah! Good catch!

**Review by rwaskiewicz — COMMENTED** [review]
One last question/comment

File: src/compiler/index.ts
Comment (alicewriteswrongs) [reply src/compiler/index.ts]: that makes sense, I will add it back 👍

**Review by rwaskiewicz — APPROVED** [review]

**Comment by alicewriteswrongs** [comment]
@tanner-reits if you wouldn't mind taking a look — I think this is waiting for your sign-off

**Review by tanner-reits — APPROVED** [review]
LGTM! Appreciate you doing some archeology and adding these comments! I'm sure they'll come in handy later

**Pull request merged**
